query/src/sql/planner/binder/aggregate.rs

```diff
@@ -12,114 +12,471 @@
+            func_name: aggregate.func_name.clone(),
+            distinct: aggregate.distinct,
+            params: aggregate.params.clone(),
+            args: replaced_args,
+            return_type: aggregate.return_type.clone(),
+        };
+
+        agg_info.aggregate_functions.push(replaced_agg.clone());
+        Ok(replaced_agg.into())
+    }
 }
 
 impl<'a> Binder {
-    /// We have supported two kinds of `group by` items:
+    /// Analyze aggregates in select clause, this will rewrite aggregate functions.
+    /// See `AggregateRewriter` for more details.
+    pub(crate) fn analyze_aggregate_select(
+        &mut self,
+        bind_context: &mut BindContext,
+        select_list: &mut SelectList<'a>,
+    ) -> Result<()> {
+        for item in select_list.items.iter_mut() {
+            let mut rewriter = AggregateRewriter::new(bind_context, &mut self.metadata);
+            let new_scalar = rewriter.visit(&item.scalar)?;
+            item.scalar = new_scalar;
+        }
+
+        Ok(())
+    }
+
+    /// Analyze aggregates in having clause, this will rewrite aggregate functions.
+    /// See `AggregateRewriter` for more details.
+    pub(super) async fn analyze_aggregate_having(
+        &mut self,
+        bind_context: &mut BindContext,
+        having: &Expr<'a>,
+    ) -> Result<Scalar> {
+        let scalar_binder = ScalarBinder::new(bind_context, self.ctx.clone());
+        let (scalar, _) = scalar_binder.bind_expr(having).await?;
+        let mut rewriter = AggregateRewriter::new(bind_context, &mut self.metadata);
+        rewriter.visit(&scalar)
+    }
+
+    /// We have supported three kinds of `group by` items:
     ///
+    ///   - Index, a integral literal, e.g. `GROUP BY 1`. It choose the 1st item in select as
+    ///     group item.
     ///   - Alias, the aliased expressions specified in `SELECT` clause, e.g. column `b` in
     ///     `SELECT a as b, COUNT(a) FROM t GROUP BY b`.
     ///   - Scalar expressions that can be evaluated in current scope(doesn't contain aliases), e.g.
     ///     column `a` and expression `a+1` in `SELECT a as b, COUNT(a) FROM t GROUP BY a, a+1`.
-    pub(crate) fn analyze_aggregate(&self, output_context: &BindContext) -> Result<AggregateInfo> {
-        let mut available_aliases: Vec<ColumnBinding> = vec![];
-        for column_binding in output_context.all_column_bindings() {
-            if let Some(scalar) = &column_binding.scalar {
-                // If a column is `BoundColumnRef`, we will treat it as an available alias anyway.
-                // Otherwise we will check visibility and if the scalar contains aggregate functions.
-                if matches!(scalar.as_ref(), Scalar::BoundColumnRef(_))
-                    || (column_binding.visible
-                        && find_aggregate_scalars(&[*scalar.clone()]).is_empty())
-                {
-                    available_aliases.push(column_binding.clone());
-                }
+    pub(super) async fn bind_aggregate(
```

Comment:
Maybe we can split the method into multi-parts, such as moving the content about filling `group_items` to a new method.

Reply:
Fixed.